Add Entry Statistics on top of QC View

samshara opened the pull request:
depends on https://github.com/the-deep/server/pull/566
fixes https://github.com/the-deep/client/issues/1183

### Note: the entry stats may change due to #1177 which is implemented by #1194. The entry stats thus changed won't be shown by the solution implemented.

samshara committed:
Add Entry Statistics on top of QC View

AdityaKhatri commented:
@samshara , I think we should pull the stats again after the lead is edited. I just merged the PR you mentioned.

samshara committed:
Add refresh stats on  entry lead change

samshara committed:
Add refresh view on entry edit

here new entries need to be pulled from the server. if we are to
patch the entry or lead information we need to apply all the filtering
logic that have been applied on the server.

samshara committed:
Add translations

tnagorra commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,65 @@
+            </div>
+        );
+    }
+    return (null);
```
Better to use `if (!stats) { return null; }` to short-circuit

tnagorra commented on src/views/Entries/QualityControl/index.tsx:
```diff
@@ -116,56 +121,86 @@ function QualityControl(props: Props) {
     );
 
-    const handleEntryDelete = React.useCallback((entryId) => {
+    const handleEntryEdit = useCallback(() => {
+        getEntries();
+    }, [getEntries]);
```
```suggestion
const handleEntryEdit = getEntires;
```

tnagorra commented on src/views/Entries/QualityControl/index.tsx:
```diff
@@ -176,15 +211,13 @@ function QualityControl(props: Props) {
-        });
-        setEntries(patchedEntries);
-    }, [entries]);
+    const handleLeadEdit = useCallback(() => {
+        getEntries();
+    }, [getEntries]);
+
+    const handleVerificationChange = useCallback(() => {
+        getEntries();
+    }, [getEntries]);
```
```suggestion
   const handleLeadEdit = getEntries;
   const handleVerificationChange = getEntries;
```

samshara committed:
Fix based on code review

frozenhelium committed:
Improve QC stats

frozenhelium committed:
Remove loading animation for entry stat load

AdityaKhatri commented on src/views/Entries/QualityControl/index.tsx:
```diff
@@ -160,6 +160,7 @@ function QualityControl(props: Props) {
         onSuccess: (response) => {
             setEntries(response.results);
             setStats(response.summary);
+            console.info(response.summary);
```
Remove this

frozenhelium replied:
We do not comment on outdated changes

frozenhelium committed:
Fix access control for QC mode

AdityaKhatri committed:
Change lead and entry in local state

AdityaKhatri commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,113 @@
+        ...staticStats
+    } = stats;
+
+    const statsList: Stats[] = Object.entries({ ...staticStats }).map(([k, v]) => ({
+        id: k,
+        title: staticEntryStatTitles[k as keyof StaticEntrySummary],
+        value: v,
+    }));
+
+    orgTypeCount.forEach((orgType) => {
+        statsList.push({
+            id: String(orgType.org.id),
+            title: orgType.org.shortName ?? orgType.org.title,
+            value: orgType.count,
+        });
+    });
+
+    const max = Math.max(...statsList.map(d => d.value || 0));
```
@frozenhelium , please memoize these calculations

tnagorra commented on src/views/Entries/QualityControl/index.tsx:
```diff
@@ -116,56 +121,112 @@ function QualityControl(props: Props) {
+    const handleEntryEdit = useCallback((updatedEntry) => {
+        setEntries(oldEntries => (
+            oldEntries.map((entry) => {
+                if (entry.id === updatedEntry.id) {
+                    return ({
+                        ...updatedEntry,
+                        lead: entry.lead,
+                    });
+                }
+                return entry;
```
Even though this works, we should use immer or splice

tnagorra commented on src/views/Entries/QualityControl/index.tsx:
```diff
@@ -116,56 +121,112 @@ function QualityControl(props: Props) {
+        getEntriesWithStats();
+        setEntries(oldEntries => (
+            oldEntries.map((entry) => {
+                if (entry.lead?.id === updatedLead.id) {
+                    return ({
+                        ...entry,
+                        lead: updatedLead,
+                    });
+                }
```
Even though this works, we should either user immer or splice

tnagorra commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,119 @@
+function EntryStat({
+    title,
+    value = 0,
+    max,
```
Should we check if max is zero? before division?

AdityaKhatri replied:
No its fine

tnagorra commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,119 @@
+            }}
+        >
+            <div className={styles.value}>
+                {value}
```
Should we not use Numeral?

tnagorra commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,119 @@
+            value: v,
+        }));
+
+        orgTypeCount.forEach((orgType) => {
+            statsList.push({
+                id: String(orgType.org.id),
+                title: orgType.org.shortName ?? orgType.org.title,
+                value: orgType.count,
+            });
+        });
```
We should use `map` instead.

tnagorra commented on src/views/Entries/QualityControl/EntriesStats/index.tsx:
```diff
@@ -0,0 +1,119 @@
+    }, [staticStats, orgTypeCount]);
+
+    const max = useMemo(() => (
+        Math.max(...statsList.map(d => d.value || 0))
```
We should include a default max value.
Also, we should prefer `??` if it is available

tnagorra committed:
Fix ordering of static stats on entries page

- Fix schema for entries
